windows/src/developer/kmcmpdll/Compiler.cpp

```diff
@@ -1504,6 +1504,40 @@ DWORD CheckStatementOffsets(PFILE_KEYBOARD fk, PFILE_GROUP gp, PWSTR context, PW
   return CERR_None;
 }
 
+/**
+ * Checks that the order of statements in the context matches the specification
+ *   Rule structure: [context] ['+' key] '>' output
+ *   Context structure: [nul] [if()|baselayout()|platform()]+ [char|any|context()|deadkey()|dk()|index()|notany()|outs()]
+ * Test that nul is first, then if(), baselayout(), platform() statements are before any other content
+ */
+BOOL CheckContextStatementPositions(PWSTR context) {
+  BOOL hadContextChar = FALSE;
+  for (PWSTR p = context; *p; p = incxstr(p)) {
+    if (*p == UC_SENTINEL) {
+      switch (*(p + 1)) {
+      case CODE_NUL:
+        if (p > context) {
+          AddWarning(CWARN_NulNotFirstStatementInContext);
+        }
+        break;
+      case CODE_IFOPT:
+      case CODE_IFSYSTEMSTORE:
+        if (hadContextChar) {
+          AddWarning(CWARN_IfShouldBeAtStartOfContext);
+        }
+        break;
+      default:
+        hadContextChar = TRUE;
+      }
+    }
+    else {
+      hadContextChar = TRUE;
+    }
+  }
+
+  return TRUE;
```

Comment:
When does `hadContextChar` get used if this always returns `TRUE`? 

I noticed l.1596 doesn't use the returned BOOL either.

Reply:
Yes, the function needs to return `BOOL` because `AddWarning` is a macro that can force an early `return(FALSE)`. `hadContextChar` is used in the `CODE_IF` test a few lines above.